File: Python/ByTier/C/1917_C_Watering_an_Array.py
# Problem: CF 1917 C - Watering an Array
# https://codeforces.com/contest/1917/problem/C

"""
Algorithm: Watering an Array
Approach:
- We maintain an array `arrn` which is derived from the original array `arr` by subtracting index + 1 from each element.
- This transformation helps in identifying elements that are equal to their position (a[j] == j) more easily.
- The idea is to simulate operations optimally, considering that we can either increment the first b[i] elements or count and reset the array.
- The key optimization is based on the periodic nature of `b` array (repeating sequence v).
- We try to simulate some increments and then compute the potential score at each step.

Time Complexity: O(d * k) in worst case, but since d can be up to 1e9 and k up to 1e5, it is optimized due to early stopping condition.
Space Complexity: O(n + k) for storing the arrays `arr`, `v`, and `arrn`.

"""

for _ in range(int(input())):
    n, k, d = map(int, input().split())
    arr = [int(i) for i in input().split()]
    v = [int(i) for i in input().split()]
    # Transform array to check for positions where element equals index
    arrn = [arr[i] - i - 1 for i in range(n)]
    # Initial potential score based on current count and remaining operations
    x = arrn.count(0) + (d - 1) // 2
    i = 1
    # While conditions allow for better score:
    while (
        min(arrn) < 0  # if there are negative values in arrn
        and i <= d - 1  # while there are remaining days
        and len([j for j in arrn if j <= 0]) + (d - i - 1) // 2 > x  # if we can do better
    ):
        # Increment first b[i] elements of arrn
        for j in range(v[(i - 1) % k]):
            arrn[j] += 1
        i += 1
        # Recalculate potential score
        y = arrn.count(0) + (d - i) // 2
        x = max(x, y)
    print(x)


# https://github.com/VaHiX/CodeForces/

File: CSharp/ByTier/C/1917_C_Watering_an_Array.cs
// Problem: CF 1917 C - Watering an Array
// https://codeforces.com/contest/1917/problem/C

/*
 * Code Purpose:
 * This program solves the "Watering an Array" problem where we have an array of integers and perform operations over 'd' days.
 * Each day we either increment the first b_i elements of the array or count how many elements equal their position and add that to the score,
 * then reset the array to zeros.
 * 
 * Algorithms/Techniques:
 * - Simulation with periodic sequence b
 * - Optimization: Only simulate a limited number of steps (2*n+1) to avoid unnecessary computation
 * - Greedy approach: For each simulation step, compute the score and track maximum
 * 
 * Time Complexity:
 * O(min(2*n + 1, d) * n) where n is the length of array a, and d is the number of days.
 * In worst case, this simplifies to O(n^2) given the constraints.
 * 
 * Space Complexity:
 * O(n + k) where n is the length of array a and k is the length of sequence v.
 * This accounts for the arrays 'entity' and 'bmps' used in the solution.
 */

using System;
using System.Linq;
using System.Collections.Generic;
class Program {
  static void Main() {
    try {
      Console.SetIn(new System.IO.StreamReader(Console.OpenStandardInput(),
                                               Console.InputEncoding, false,
                                               bufferSize: 1024));
      Console.SetOut(new System.IO.StreamWriter(Console.OpenStandardOutput(),
                                                Console.OutputEncoding,
                                                bufferSize: 1024));
      int blind = int.Parse(Console.ReadLine());
      for (int testCase = 0; testCase < blind; testCase++) {
        int soul, gdr, insane;
        soul = gdr = insane = 0;
        string[] input = Console.ReadLine().Split();
        soul = int.Parse(input[0]); // Length of array a
        gdr = int.Parse(input[1]); // Length of sequence v
        insane = int.Parse(input[2]); // Number of days
        int[] entity = new int[soul + 1]; // Array a (1-indexed)
        int[] bmps = new int[gdr + 1]; // Sequence v (1-indexed)
        string[] entityInput = Console.ReadLine().Split();
        for (int i = 1; i <= soul; i++) {
          entity[i] = int.Parse(entityInput[i - 1]); // Read array a values
        }
        string[] bmpsInput = Console.ReadLine().Split();
        for (int i = 1; i <= gdr; i++) {
          bmps[i] = int.Parse(bmpsInput[i - 1]); // Read sequence v values
        }
        int testCnt = Math.Min(2 * soul + 1, insane); // Limit simulation steps to avoid TLE
        int ans = 0, idx = 1; // ans stores max score, idx is pointer to current element in sequence v
        for (int toZero = 1; toZero <= testCnt; toZero++) {
          int cur = 0;
          // Count how many elements equal their position
          for (int i = 1; i <= soul; i++) {
            if (entity[i] == i)
              cur++;
          }
          // Estimate potential future score (optimization heuristic)
          cur += (insane - toZero) / 2;
          ans = Math.Max(ans, cur); // Update maximum score
          // Increment the first bmps[idx] elements
          for (int i = 1; i <= bmps[idx]; i++)
            entity[i]++;
          idx++; // Move to next element in sequence v (wraps around if needed)
          if (idx > gdr)
            idx = 1;
        }
        Console.WriteLine(ans);
      }
      Console.Out.Flush();
    } catch (Exception ex) {
      Console.WriteLine($"An error occurred: {ex.Message}");
    }
  }
}


// https://github.com/VaHiX/CodeForces/